mbed/mbed.py

```diff
@@ -2497,10 +2497,24 @@ def compile_(toolchain=None, target=None, profile=False, compile_library=False,
 
         build_path = build
 
+        def safe_append_profile_to_build_path(bp):
```

Comment:
Please avoid functions within functions. This makes it much more difficult to track what state is used by/shared with this function.

Reply:
I nested it because it's a helper function with limited scope, and likely not useful elsewhere.

Is your issue the fact that it's using a variable (`profile`) from its outer scope? Would adding that as a parameter be sufficient?

Or is there some other issue I'm missing? Or is it just a rule that you want to be strictly followed, because of the *potential* for abuse, i.e. using variables from outer scope?

Reply:
Yes, my preference is to avoid using closures. Not because of their potential for abuse, but because it increases the amount of code I have to read to understand what state affects a single function. A function on it's own can change behavior based on global variable and inputs. A closure, a function within the body of another function, adds that function's scope and variable into the pool of variables that may impact it's state.

Helper functions may be prefixed with an `_` to mark them private, by python convention.

TL;DR: I find self-contained functions easier to read and understand.

Reply:
It also might be useful to make the same change to the `mbed test` sub-command.